feat: Puck adheres to safe area as window resizes

shirakaba opened the pull request:
As the window resizes, the puck will stay in the safe area. For desktop, the safe area insets are all `0px` in practice, so it's not too exciting (but it's great that now the puck will not fall out of the viewport).

On mobile, it'll ensure that you can't move the puck beyond the safe area (i.e. into the notch), and will keep it onscreen despite device rotations.

https://user-images.githubusercontent.com/14055146/124359389-9c099200-dc1c-11eb-8080-34afb5e032f9.mp4

shirakaba commented:
And on an iPhone 11 Pro running the iOS 15 beta. Note that I've been finding it perhaps harder to move the puck now that we're using pointermove rather than mousemove – frequently the page scrolls, or the text starts getting selected. I'm unsure whether moving to pointer up/down will improve things in this area, but it'll be interesting to see.

https://user-images.githubusercontent.com/14055146/124360145-1f78b280-dc20-11eb-8520-c354dffe0009.mp4

birtles reviewed (changes requested):
This looks really really great. The videos are also very nice!

I made a few suggestions though. I can probably fix them if you like?

shirakaba commented on src/puck.ts:
```diff
@@ -22,6 +22,55 @@ export class RikaiPuck {
+      isNaN(safeAreaLeft)
+    ) {
+      return;
+    }
```
Done!

shirakaba commented on src/puck.ts:
```diff
@@ -22,6 +22,55 @@ export class RikaiPuck {
+     * @see https://developer.mozilla.org/en-US/docs/Web/API/Window/innerWidth#usage_notes
+     * @see https://stackoverflow.com/a/7205786/5951226
+     */
+    const { innerWidth, innerHeight } = window;
```
> We should use `clientWidth` here as even modern browsers don't overlay scrollbars on Windows. It's also consistent with what we do for popup positioning:

I went back and forth on this one, so perfectly happy to settle on `clientWidth` for consistency with the popup. I am unclear whether we have to do any extra maths to compensate for hypothetical interaction between a (non-zero) safe area and (non-overlaid) scroll bars, but I don't believe a device with that combination exists anyway. I've implemented something to address this now.

> It would be better if we cache this value after we first read it and invalidate the cached value on each `resize` or `deviceorientation` event.

Good point, I've implemented something to address that now. One difficulty, though: on iOS Safari, the safe area can very much update without a window resize. Whether it emits a "resize" event at that moment is worth investigating, but either way – if you tap into the URL bar, it'll enlarge it, thus adjusting the safe area, and I'm not sure whether we have any way to listen for safe area changes like that. We may just have to recalculate it on all pointermove events emitted while the Puck is being dragged.

Side note: I once came across a blog post investigating which CSS properties caused flushing upon being read, but I've long since lost it. Might've been a David Walsh or similar; I don't suppose you have a resource for which properties cause style and layout flushing?

> (It's possible all browsers now dispatch both resize and deviceorientation when you rotate the device and we don't actually need to listen for deviceorientation events but I think there was a point where at least one browser didn't, and I think it was Safari.)

I can confirm that iOS Safari emits a "resize" event upon rotate (as that's what we rely upon in the video recording above where the puck stays within the safe area throughout rotations).

birtles replied:
> Side note: I once came across a blog post investigating which CSS properties caused flushing upon being read, but I've long since lost it. Might've been a David Walsh or similar; I don't suppose you have a resource for which properties cause style and layout flushing?

I'm aware of https://csstriggers.com/ but it only tells you what _changing_ the value does.

Paul Irish has a list too but it's fairly Chrome specific: https://gist.github.com/paulirish/5d52fb081b3570c81e3a. It's also not quite accurate since `getComputedStyle(elem)` doesn't actually trigger anything in any browser. It's only when you _read_ one of the properties from the returned object that things may be flushed.

`getComputedStyle(elem).width` will flush layout, while typically `getComputedStyle(elem).backgroundColor` will only trigger style to be flushed (that used to be only true in Firefox but it looks like Chrome now has that optimization).

Other than that, anything which depends on the dimensions of elements / the page being known will generally flush layout hence why `window.innerWidth` etc. flush layout.

(There are some other odd cases too like `transform` which technically doesn't need to flush layout since it is applied post-layout, but which occasionally does in order to update scrollbars. e.g. in Firefox when there is a transform animation, it flushes layout every 300ms to keep the scrollbars in sync.)

It's not a big problem in this case but we're being injected into a Web page which could be doing anything. If we trigger layout to be flushed on every `mousemove` event we could end up causing layout thrashing or even triggering transitions at the wrong time (since triggering a style flush triggers transitions) so we should be a bit careful.

shirakaba commented:
@birtles thanks for the review – very important points raised with respect to performance.

I'm having difficulties with switching from `window.innerHeight` to `document.body.clientHeight`, however. The following screenshots are taken when using `document.body.clientHeight`:

# Portrait orientation

## Minimum safe area

<table>
<tr>
<td>
Puck top limit
</td>
<td>
Puck bottom limit
</td>
</tr>
<tr>
<td>
<img width="300" alt="image" src="https://user-images.githubusercontent.com/14055146/124673239-6a105e00-deb0-11eb-9289-b826fd3f911e.png">
</td>
<td>
<img width="300" alt="image" src="https://user-images.githubusercontent.com/14055146/124674404-91682a80-deb2-11eb-97e6-2f37aa28fea3.png">
</td>
</tr>
</table>

This code snippet:

```js
{
    windowInnerWidth: window.innerWidth,
    windowInnerHeight: window.innerHeight,
    windowOuterWidth: window.outerWidth,
    windowOuterHeight: window.outerHeight,
    documentBodyClientWidth: document.body.clientWidth,
    documentBodyClientHeight: document.body.clientHeight,
    safeAreaTop: getComputedStyle(document.getElementById("tenten-ja-puck").shadowRoot.lastElementChild).getPropertyValue("--tenten-puck-safe-area-inset-top"),
    safeAreaRight: getComputedStyle(document.getElementById("tenten-ja-puck").shadowRoot.lastElementChild).getPropertyValue("--tenten-puck-safe-area-inset-right"),
    safeAreaBottom: getComputedStyle(document.getElementById("tenten-ja-puck").shadowRoot.lastElementChild).getPropertyValue("--tenten-puck-safe-area-inset-bottom"),
    safeAreaLeft: getComputedStyle(document.getElementById("tenten-ja-puck").shadowRoot.lastElementChild).getPropertyValue("--tenten-puck-safe-area-inset-left"),
}
```

... evaluates as:

```js
{
    windowInnerWidth: 375,
    windowInnerHeight: 768,
    windowOuterWidth: 375,
    windowOuterHeight: 812,
    documentBodyClientWidth: 375,
    documentBodyClientHeight: 716,
    safeAreaTop: " 0px",
    safeAreaRight: " 0px",
    safeAreaBottom: " 112px",
    safeAreaLeft: " 0px"
}
```

##  Maximum safe area

<table>
<tr>
<td>
Puck top limit
</td>
<td>
Puck bottom limit
</td>
</tr>
<tr>
<td>
<img width="300" alt="image" src="https://user-images.githubusercontent.com/14055146/124673610-12bebd80-deb1-11eb-81f0-b07a9e9f98a3.png">
</td>
<td>
<img width="300" alt="image" src="https://user-images.githubusercontent.com/14055146/124674346-70073e80-deb2-11eb-8170-8cf6423337f0.png">
</td>
</tr>
</table>

The same code snippet evaluates as:

```diff
{
    windowInnerWidth: 375,
-   windowInnerHeight: 768,
+   windowInnerHeight: 716,
    windowOuterWidth: 375,
    windowOuterHeight: 812,
    documentBodyClientWidth: 375,
    documentBodyClientHeight: 716,
    safeAreaTop: " 0px",
    safeAreaRight: " 0px",
-   safeAreaBottom: " 112px",
+   safeAreaBottom: " 0px",
    safeAreaLeft: " 0px"
}
```

birtles commented:
> **EDIT:** I realise that perhaps I should strictly have been looking at `document.body.parentElement.clientHeight` instead, but it looks like on iOS Safari, it has identical values to `document.body.clientHeight` anyway. It may be different for browsers with non-overlaid scrollbars. But either way, we have a test failure here.

Right, I think in standards mode `document.documentElement.clientHeight` gives you the viewport height but `document.body.clientHeight` gives you the document height. But in quirks mode it's different:

https://github.com/birchill/10ten-ja-reader/blob/33922b3e02ae9bdbb9ff8a2e1119d6b9af83acfa/src/popup-position.ts#L37-L40

> Should we revert to `window.innerWidth` and `window.innerHeight`, or can you see another approach that would keep it viable?

If `window.innerHeight` is proving difficult I think it would be totally fine to use `document.documentElement.clientWidth` and `window.innerHeight`. Horizontal scrollbars are infrequent enough that we don't need to worry about them too much, I think (I mean, they can be common on mobile devices, but in those cases they tend to be overlaid scrollbars).

birtles reviewed (commented):
I left a few comments but everything else looks great! Thank you!

shirakaba commented:
I've addressed those comments.

I think I've explained the mysterious 52 px change in `clientHeight` before and after entering iOS 15 Safari's "compact mode":

<img width="1354" alt="image" src="https://user-images.githubusercontent.com/14055146/124731481-2cd8ba00-df0a-11eb-84ae-d0edd6567bd4.png">

Gotta dash!

birtles commented:
This looks really great. Thank you so much. Let's merge this in.
